openstack_dashboard/api/glance.py

```diff
@@ -152,64 +154,42 @@ def __ne__(self, other_image):
+            if not Image.PUBLISHED_APPLIANCES.get(self.id):
+                return -1
+
+            return Image.PUBLISHED_APPLIANCES.get(self.id).get('id')
         except:
             LOG.error('Error getting catalog id from appliance catalog api for image id: ' + self.id)
         return -1
 
     def get_is_project_supported(self):
         try:
-            LOG.info('Getting project_supported flag from appliance catalog api for image id: ' + self.id)
-            app_json = json.loads(Image.APP_CATALOG)
-            for app in app_json['result']:
-                if(app['chi_uc_appliance_id'] == self.id):
-                    return app['project_supported']
-                if(app['chi_tacc_appliance_id'] == self.id):
-                    return app['project_supported']
-                if(app['kvm_tacc_appliance_id'] == self.id):
-                    return app['project_supported']
+            if not Image.PUBLISHED_APPLIANCES.get(self.id):
+                return False
+
+            return Image.PUBLISHED_APPLIANCES.get(self.id).get('project_supported')
         except:
             LOG.error('Error getting project_supported flag from appliance catalog api for image id: ' + self.id)
         return False
 
     def get_is_published_in_app_catalog(self):
         try:
-            LOG.info('Checking if image is published to appliance catalog for image id: ' + self.id)
-            app_json = json.loads(Image.APP_CATALOG)
-            for app in app_json['result']:
-                if(app['chi_uc_appliance_id'] == self.id):
-                    return True
-                if(app['chi_tacc_appliance_id'] == self.id):
-                    return True
-                if(app['kvm_tacc_appliance_id'] == self.id):
-                    return True
+            return Image.PUBLISHED_APPLIANCES.get(self.id) is not None
         except:
-            LOG.error('Error checking if image is published to appliance catalog for image id: ' + self.id)
+             LOG.error('Error checking if image is published to appliance catalog for image id: ' + self.id)
         return False
 
-@memoized
-def fetch_supported_appliances(request):
-    try:
-        LOG.info('Fetching and caching Appliance JSON from ' + settings.CHAMELEON_PORTAL_API_BASE_URL + settings.APPLIANCE_CATALOG_API_PATH)
-        http = urllib3.PoolManager()
-        r = http.request('GET', settings.CHAMELEON_PORTAL_API_BASE_URL + settings.APPLIANCE_CATALOG_API_PATH)
-        LOG.debug('fetched appliance catalog data: ' + r.data)
-        return r.data
-    except Exception as e:
-        LOG.error(e)
-    return []
+def fetch_published_appliances(request):
+    if not cache.get('app_catalog_updated'):
+        try:
+            LOG.info('Fetching appliances from ' + settings.CHAMELEON_PORTAL_API_BASE_URL + settings.APPLIANCE_CATALOG_API_PATH)
+            Image.PUBLISHED_APPLIANCES = requests.get(settings.CHAMELEON_PORTAL_API_BASE_URL + settings.APPLIANCE_CATALOG_API_PATH).json()
```

Comment:
Is `Image.PUBLISHED_APPLIANCES` going to be shared across threads? Why not store the response from Portal directly in the cache and pull from it instead of fetching from the remote API if the cache is filled?

Reply:
Yes,  `Image.PUBLISHED_APPLIANCES` is shared across threads here, though if I stored the response in the cache it would also be shared across threads, in this case I'm using the cache as a marker for the refresh interval. I use the static var to keep published appliance data available to the front-end during that (hopefully brief) period between the time the cache expires and when new data is pulled down and updated.